### PR TITLE
CLI - separate build codegen actions

### LINKED ISSUE #1365: CLI - Separate `build` and `codegen` command actions
Related to #1363 

When running the `polywrap codegen` command within a Plugin project, an additional build step is executed.

This logic/step should be moved into the `build` command to further improve semantics between `codegen` and `build`.

### PR DESCRIPTION
**Pull request by pileks**
Closes #1365 and allows for further work on #1363 and others (namely Plugin Build Strategies, TBD).

~~**CLI test cases will be fixed once #1384 is merged!**~~ Merged.

This PR entirely separates the `build` and `codegen` command functionalities.

Running `polywrap codegen` on a `plugin` project will no longer emit `wrap.info` into the `./build` folder.

**This PR also introduces a possible circular dependency issue, as some of our plugin projects will stop working once the newest CLI version is released.**
A possible workaround for this is to have the `build` script of each project run `polywrap codegen` and continue with the build regardless of the result of the codegen, then removing it once we have the next release up.
Thoughts?

Remaining work/merge prerequisites:

- [x] Merge #1384 
- [x] See what to do in regards to the circular dependency issue described above.
  - I have decided to solve this by adding a `build:patch` script to all plugin packages which attempts to run `polywrap build` in order to output the WRAP ABI. If it fails, a message is `echo`ed and the script exits with code 0, making the build work for the current version `polywrap` in `dependencies` and for the new, unreleased version. This way we won't have an interruption in the build process, and can later on write out the proper codegen/build logic (which is first `codegen` then `build` AND make sure that both exit succesfully).

### CODE REVIEW

**Review by Niraj-Kamdar — DISMISSED** [review]
LGTM!

File: packages/cli/src/commands/build.ts [hunk 1]
```diff
@@ -164,51 +186,123 @@ async function run(options: BuildCommandOptions) {
...
     const result = await execute();
 
     if (!result) {
       process.exit(1);
     }
+
+    process.exit(0);
   } else {
     // Execute
     await execute();
```
Comment (krisbitney) [inline packages/cli/src/commands/build.ts:262]: Is `execute` called twice in this branch? On line 308 and in the watcher?

File: packages/cli/src/commands/build.ts [hunk 2]
```diff
@@ -164,51 +186,123 @@ async function run(options: BuildCommandOptions) {
...
-    polywrapManifestPath: manifestFile,
-    logger,
-  });
-  await project.validate();
+  const project = await getProjectFromManifest(manifestFile, logger);
 
-  const polywrapManifest = await project.getManifest();
-  await validateManifestModules(polywrapManifest);
+  if (!project) {
+    return;
```
Comment (krisbitney) [inline packages/cli/src/commands/build.ts:188]: Should we return an exit code here?

**Review by krisbitney — CHANGES_REQUESTED** [review]
LGTM! 

I left a couple of comments with questions. Let me know what you think. Maybe no changes are needed.

File: packages/cli/src/commands/build.ts (hunk 2 above)
Comment (pileks) [reply packages/cli/src/commands/build.ts:188]: ...and print an error, yes!

File: packages/cli/src/commands/build.ts (hunk 1 above)
Comment (pileks) [reply packages/cli/src/commands/build.ts:262]: Yes it is!

We execute once first, then set up the watcher, as the watcher only triggers on changes in watched files.

Comment (krisbitney) [reply packages/cli/src/commands/build.ts:262]: Okay, right, understood!

**Review by krisbitney — DISMISSED** [review]

**Comment by krisbitney** [comment]
Can this be merged @pileks @dOrgJelli @Niraj-Kamdar ?

**Comment by pileks** [comment]
@krisbitney From a standpoint of implementation - Yes, it's just a matter of being reviewed! 😄

File: packages/cli/lang/en.json [hunk 3]
```diff
@@ -17,11 +18,13 @@
...
   "commands_build_options_options": "options",
   "commands_build_options_t": "Use the development server's ENS instance",
-  "commands_build_options_n": "Skip code generation",
+  "commands_build_options_codegen": "Perform code generation before build",
+  "commands_build_options_codegen_dir": "Codegen output directory (default: {default})",
   "commands_build_options_w": "Automatically rebuild when changes are made (default: false)",
   "commands_build_options_s": "Strategy to use for building the wrapper",
   "commands_build_options_s_strategy": "strategy",
   "commands_build_options_l": "Log file to save console output to",
+  "commands_build_error_codegen_failed": "Code generation failed!",
```
Comment (dOrgJelli) [inline packages/cli/lang/en.json]: NIT: I don't think we need to add exclamations (`!`) to output messages, I don't believe we do this elsewhere.

**Review by krisbitney — DISMISSED** [review]

File: packages/cli/src/commands/build.ts [hunk 4]
```diff
@@ -164,51 +186,123 @@ async function run(options: BuildCommandOptions) {
...
-  };
+      const result = await compiler.compile();
+
+      if (!result) {
+        return result;
+      }
+
+      return true;
+    };
+  } else if (isPluginManifestLanguage(language)) {
```
Comment (dOrgJelli) [inline packages/cli/src/commands/build.ts]: I think that we should be handling this inside of the "Compiler", like we already do for interface projects. I'll make a push now to add this.

**Review by dOrgJelli — CHANGES_REQUESTED** [review]

File: packages/cli/lang/en.json (hunk 3 above)
Comment (dOrgJelli) [reply packages/cli/lang/en.json]: Made these changes here: https://github.com/polywrap/toolchain/pull/1467

File: packages/cli/src/commands/build.ts (hunk 4 above)
Comment (dOrgJelli) [reply packages/cli/src/commands/build.ts]: Made these changes here: https://github.com/polywrap/toolchain/pull/1467

**Review by dOrgJelli — CHANGES_REQUESTED** [review]
Also @pileks we should update the CLI's README.

File: packages/cli/lang/en.json (hunk 3 above)
Comment (pileks) [reply packages/cli/lang/en.json]: Thanks for the PR, merged.

**Comment by pileks** [comment]
@dOrgJelli I've performed the minimal required changes to the readme just now.

We can refactor the readme to use `doc-snippets` in a later PR - for now this should be it.

LMK if this works.

**Review by dOrgJelli — APPROVED** [review]

**Pull request merged**